pubtools/_pulp/tasks/push/items/base.py

```diff
@@ -281,11 +300,42 @@ def refresh_after_copy(_):
 
             return f
 
+        # A helper function to re-try copying of failed items
+        #  Since the association phase can take a while and the items can be
+        #  moved to other repos meanwhile, we fetch up-to-date info about the
+        #  items from pulp and re-try the failing cases
+        def retry_failed_items():
+            # Get an up-to-date version of all the copy items.
+            f = cls.items_with_pulp_state_single_batch(pulp_client, copy_items)
+
+            copy_results = []
+            updated_items = list(f.result())
+            if len(updated_items) > 0:
+                # copy any items
+                _, _, copy_results = enqueue_items(updated_items)
+
+            failure_count = len(copy_results)
+            # Return a future with list of all results of copy operations
+            return f_sequence(copy_results), failure_count
+
         # This future completes once *all* copies are done successfully.
         # TODO: this still could be improved, as not every item needs every copy
         # before the state could be refreshed.
         all_copies = f_sequence(copy_results)
 
+        # If there are any failed copy operations, retry them
+        for i in range(association_retry_count):
+            if len(all_copies.result()) > 0:
```

Comment:
The call to `.result()` here will cause the whole method to become blocking, which is probably going to slow down large pushes a lot.

The caller of this (Associate phase) looks like this:

```
    def run(self):
        for batch in self.iter_for_associate():
            for items in PulpPushItem.items_by_type(batch):
                for associated_f in PulpPushItem.associated_items_single_batch(
                    self.pulp_client, items, self.copy_options
                ):
                    self.put_future_outputs(associated_f)
```

Prior to your change the behavior was:
- associated_items_single_batch starts copy of a batch and returns a future.
- Associate arranges for the outputs to be sent to the next phase when that future completes.

Since this loop only starts for (and doesn't await) completion of batches, it can start processing of more than one batch at once.

Now with your change the behavior would be:

- associated_items_single_batch starts copy of a batch, *waits for it to complete* (because `.result()` was called) and returns a future.
- Associate arranges for the outputs to be sent to the next phase when that future completes (which is always immediately).

That changes the loop so that it's only able to process a single batch (of default size 1,000 items) at a time.  I checked a completed large push at https://url.corp.redhat.com/x and Associate was working on up to 29,000 items at once, so I think if there is a change which causes it to only handle 1,000 items at once there will be a significant performance impact.

I think this needs to be reworked, the retry has to happen *inside* of the future monad rather than awaiting completion of the future.  The way you had it earlier seemed more correct in this regard (but had the problems of not logging & not checking if we actually had to retry).

Reply:
Thanks for the explanation! I have completely missed that. 

I have reworked the PR, so that the retry operation is not blocking.